include/flux/op/cartesian_product.hpp

```diff
@@ -216,6 +246,37 @@ struct sequence_traits<detail::cartesian_product_adaptor<Bases...>>
         }(std::index_sequence_for<Bases...>{});
     }
 
+    template <std::size_t I, typename Self, typename Function,
+              typename... PartialElements>
+    static constexpr auto for_each_while_impl(Self& self,
+                                              Function&& func,
+                                              PartialElements&&... partial_elements)
+        -> std::tuple<bool, detail::cartesian_product_partial_cursor_t<Self, I>>
+    {
+        // We need to iterate right to left.
+        if constexpr (I == sizeof...(Bases) - 1) {
+            bool keep_going = true;
+            auto this_current = flux::for_each_while(std::get<I>(self.bases_),
+                [&](auto&& elem) {
+                    keep_going = std::invoke(func,
+                        element_t<Self>(FLUX_FWD(partial_elements)..., FLUX_FWD(elem)));
+                    return keep_going;
+                });
+            return std::tuple(keep_going, std::tuple(std::move(this_current)));
+        } else {
+            bool keep_going = true;
+            detail::cartesian_product_partial_cursor_t<Self, I+1> nested_current;
+            auto this_current = flux::for_each_while(std::get<I>(self.bases_),
+                [&](auto&& elem) {
+                    std::tie(keep_going, nested_current) = for_each_while_impl<I+1>(
+                        self, func, FLUX_FWD(partial_elements)..., FLUX_FWD(elem));
+                    return keep_going;
+                });
+            return std::tuple(keep_going,
+                std::tuple_cat(std::tuple(std::move(this_current)), std::move(nested_current)));
+        }
+    }
```

Comment:
Thinking out loud: rather than returning a `(keep_going, partial_cursor)` tuple, could `for_each_while_impl` take and pass a `bool&` out parameter instead, and just return the partial cursor? 

This would save having to do a `tuple_cat` in the return, which from memory is quite expensive compile-time wise, and might also save needing a local variable at every level?

Reply:
Done in abba492d018d7801c150117e4a320854a1a2b69b. I had to pass the cursor as a reference parameter as well. This removes the need for the partial cursor structure entirely.